packages/react-reconciler/src/ReactUpdateQueue.js

```diff
@@ -0,0 +1,639 @@
+) {
+  // Append the update to the end of the list.
+  if (queue.lastUpdate === null) {
+    // Queue is empty
+    queue.firstUpdate = queue.lastUpdate = update;
+  } else {
+    queue.lastUpdate.next = update;
+    queue.lastUpdate = update;
+  }
+  if (
+    queue.expirationTime === NoWork ||
+    queue.expirationTime > expirationTime
+  ) {
+    // The incoming update has the earliest expiration of any update in the
+    // queue. Update the queue's expiration time.
+    queue.expirationTime = expirationTime;
+  }
+}
+
+export function enqueueUpdate<State>(
+  fiber: Fiber,
+  update: Update<State>,
+  expirationTime: ExpirationTime,
+) {
+  // Update queues are created lazily.
+  const alternate = fiber.alternate;
+  let queue1;
+  let queue2;
+  if (alternate === null) {
+    // There's only one fiber.
+    queue1 = fiber.updateQueue;
+    queue2 = null;
+    if (queue1 === null) {
+      queue1 = fiber.updateQueue = createUpdateQueue(fiber.memoizedState);
+    }
+  } else {
+    // There are two owners.
+    queue1 = fiber.updateQueue;
+    queue2 = alternate.updateQueue;
+    if (queue1 === null) {
+      if (queue2 === null) {
+        // Neither fiber has an update queue. Create new ones.
+        queue1 = fiber.updateQueue = createUpdateQueue(fiber.memoizedState);
+        queue2 = alternate.updateQueue = createUpdateQueue(
+          alternate.memoizedState,
+        );
+      } else {
+        // Only one fiber has an update queue. Clone to create a new one.
+        queue1 = fiber.updateQueue = cloneUpdateQueue(queue2);
+      }
+    } else {
+      if (queue2 === null) {
+        // Only one fiber has an update queue. Clone to create a new one.
+        queue2 = alternate.updateQueue = cloneUpdateQueue(queue1);
+      } else {
+        // Both owners have an update queue.
+      }
+    }
+  }
+  if (queue2 === null || queue1 === queue2) {
+    // There's only a single queue.
+    appendUpdateToQueue(queue1, update, expirationTime);
+  } else {
+    // There are two queues. We need to append the update to both queues,
+    // while accounting for the persistent structure of the list — we don't
+    // want the same update to be added multiple times.
+    if (queue1.lastUpdate === null || queue2.lastUpdate === null) {
+      // One of the queues is not empty. We must add the update to both queues.
+      appendUpdateToQueue(queue1, update, expirationTime);
+      appendUpdateToQueue(queue2, update, expirationTime);
+    } else {
+      // Both queues are non-empty. The last update is the same in both lists,
+      // because of structural sharing. So, only append to one of the lists.
+      appendUpdateToQueue(queue1, update, expirationTime);
+      // But we still need to update the `lastUpdate` pointer of queue2.
+      queue2.lastUpdate = update;
+    }
+  }
+
+  if (__DEV__) {
+    if (
+      fiber.tag === ClassComponent &&
+      (queue1.isProcessing || (queue2 !== null && queue2.isProcessing)) &&
```

Comment:
Typically we just use current owner or debug frame or some global state to determine if we’re in a pure callback.

Can’t we do that with isProcessing too? We should keep state as broad as possible so we don’t have to repeat the storage n time.

Reply:
It’s also easy to mess up and try to access it in prod or define it in prod.

Reply:
We already have an `debugRenderPhaseSideEffects` check outside of a DEV block: https://github.com/facebook/react/blob/5dcf93d146ccde90af2c442d6cb32b29efddb46b/packages/react-reconciler/src/ReactFiberUpdateQueue.js#L297-L305

because `debugRenderPhaseSideEffects` is only true in DEV mode. But I agree we should put it in a DEV check anyway, for extra safety.